// @potion/element/src/Polyline.js
import PropTypes from 'prop-types';
import { line } from 'd3-shape';
import itsSet from 'its-set';

import Element from './Element';

export default class Area extends Element {

  static displayName = 'Area';

  static propTypes = {
    x: PropTypes.oneOfType([PropTypes.number, PropTypes.func]),
    y: PropTypes.oneOfType([PropTypes.number, PropTypes.func]),
    defined: PropTypes.func,
    curve: PropTypes.func,
    points: PropTypes.array.isRequired,
  };

  static defaultProps = {
    ...Element.defaultProps,
  };

  defaultComponent='path'

  getSchema() {
    return {
      d: {
        get inputs() {
          return [
            'x',
            'y',
            'defined',
            'curve',
          ];
        },
        calculation(props) {
          let calc = line();
          const keys = this.inputs;
          keys.forEach(attrName => {
            if (itsSet(props[attrName])) {
              calc = calc[attrName](props[attrName]);
            }
          });
          return calc(props.points);
        },
      },
    };
  }

}
